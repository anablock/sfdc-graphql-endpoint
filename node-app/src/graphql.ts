import {
    GraphQLBoolean,
    GraphQLEnumType,
    GraphQLFieldConfig,
    GraphQLFieldResolver,
    GraphQLFloat,
    GraphQLID,
    GraphQLInputFieldConfig,
    GraphQLInputObjectType,
    GraphQLInt,
    GraphQLList,
    GraphQLNonNull,
    GraphQLObjectType,
    GraphQLOutputType,
    GraphQLScalarType,
    GraphQLSchema,
    GraphQLString,
} from 'graphql';

import {
    Entity,
    Field,
    FieldType,
    isPolymorphicReference,
    isReferenceField,
    isScalarField,
    ReferenceField,
    ScalarField,
} from './entity.js';

declare module 'graphql' {
    interface GraphQLObjectTypeExtensions {
        sfdc?: Entity;
    }

    interface GraphQLFieldExtensions<_TSource, _TContext, _TArgs = any> {
        sfdc?: Field;
    }
}

interface SchemaConfig<Context> {
    entities: Entity[];
    resolvers?: {
        query?: (entity: Entity) => GraphQLFieldResolver<unknown, Context>;
        queryMany?: (entity: Entity) => GraphQLFieldResolver<unknown, Context>;
    };
}

interface State {
    config: SchemaConfig<any>;
    types: Map<string, GraphQLObjectType>;
    columnExprTypes: Map<string, GraphQLInputObjectType>;
    orderByTypes: Map<string, GraphQLList<GraphQLInputObjectType>>;
}

const SCALAR_TYPES_MAPPING: { [type in ScalarField['type']]: GraphQLScalarType } = {
    // Builtin GraphQL scalars
    [FieldType.ID]: GraphQLID,
    [FieldType.STRING]: GraphQLString,
    [FieldType.BOOLEAN]: GraphQLBoolean,
    [FieldType.INT]: GraphQLInt,
    [FieldType.FLOAT]: GraphQLFloat,

    // Custom SFDC types
    [FieldType.DATE]: new GraphQLScalarType({ name: 'Date' }),
    [FieldType.DATETIME]: new GraphQLScalarType({ name: 'DateTime' }),
    [FieldType.BASE64]: new GraphQLScalarType({ name: 'Base64' }),
    [FieldType.CURRENCY]: new GraphQLScalarType({ name: 'Currency' }),
    [FieldType.TEXTAREA]: new GraphQLScalarType({ name: 'TextArea' }),
    [FieldType.PERCENT]: new GraphQLScalarType({ name: 'Percent' }),
    [FieldType.PHONE]: new GraphQLScalarType({ name: 'Phone' }),
    [FieldType.URL]: new GraphQLScalarType({ name: 'URL' }),
    [FieldType.EMAIL]: new GraphQLScalarType({ name: 'Email' }),
    [FieldType.COMBOBOX]: new GraphQLScalarType({ name: 'Combobox' }),
    [FieldType.ANY_TYPE]: new GraphQLScalarType({ name: 'AnyType' }),
    [FieldType.PICKLIST]: new GraphQLScalarType({ name: 'Picklist' }),
    [FieldType.MULTI_PICKLIST]: new GraphQLScalarType({ name: 'MultiPicklist' }),
    [FieldType.ADDRESS]: new GraphQLScalarType({ name: 'Address' }),
    [FieldType.LOCATION]: new GraphQLScalarType({ name: 'Location' }),
};

const OPERATOR_INPUT_TYPES_MAPPING = Object.fromEntries(
    Object.entries(SCALAR_TYPES_MAPPING).map(([key, type]) => [key, createInputOperator({ type })]),
) as { [type in ScalarField['type']]: GraphQLInputObjectType };

const BY_ID_INPUT_ARGS = {
    id: {
        type: GraphQLID,
    },
};

const PAGINATION_INPUT_ARGS = {
    limit: {
        type: new GraphQLNonNull(GraphQLInt),
    },
    offset: {
        type: GraphQLInt,
    },
};

export type GraphQLSortOrderValue =
    | 'ASC'
    | 'DESC'
    | 'ASC_NULLS_FIRST'
    | 'ASC_NULLS_LAST'
    | 'DESC_NULLS_FIRST'
    | 'DESC_NULLS_LAST';

const ORDER_BY_ENUM_TYPE_VALUES: { [name in GraphQLSortOrderValue]: {} } = {
    ASC: {},
    DESC: {},
    ASC_NULLS_FIRST: {},
    ASC_NULLS_LAST: {},
    DESC_NULLS_FIRST: {},
    DESC_NULLS_LAST: {},
};

const ORDER_BY_ENUM_TYPE = new GraphQLEnumType({
    name: 'OrderByEnum',
    values: ORDER_BY_ENUM_TYPE_VALUES,
});

export function entitiesToSchema<Context>(config: SchemaConfig<Context>): GraphQLSchema {
    const state: State = {
        config,
        types: new Map(),
        orderByTypes: new Map(),
        columnExprTypes: new Map(),
    };

    for (const entity of config.entities) {
        const type = createGraphQLEntityType(state, entity);
        state.types.set(entity.sfdcName, type);
    }

    const query = createQuery(state, config.entities);

    return new GraphQLSchema({
        query,
        types: Array.from(state.types.values()),
    });
}

function createGraphQLEntityType(state: State, entity: Entity): GraphQLObjectType {
    const { gqlName, fields } = entity;

    return new GraphQLObjectType({
        name: gqlName,
        fields: () => {
            return Object.fromEntries(
                fields.map((field) => [field.gqlName, createGraphQLEntityField(state, field)]),
            );
        },
        extensions: {
            sfdc: entity,
        },
    });
}

function createGraphQLEntityField(state: State, field: Field): GraphQLFieldConfig<any, unknown> {
    let type: GraphQLOutputType;
    let resolve: GraphQLFieldResolver<any, unknown> = (source) => {
        return source[field.sfdcName];
    };

    if (isScalarField(field)) {
        type = SCALAR_TYPES_MAPPING[field.type];
    } else if (isReferenceField(field)) {
        // If the the referenced entity is part of the graph, make the field resolve to the
        // associated GraphQL type. Otherwise make the field resolve object id.
        if (state.types.has(field.sfdcReferencedEntityName)) {
            type = state.types.get(field.sfdcReferencedEntityName)!;
            resolve = (source) => {
                return source[field.sfdcReferencedEntityName];
            };
        } else {
            // TODO: Add support for lookup that aren't part of the graph.
            type = GraphQLID;
        }
    } else {
        // TODO: Add support for polymorphic field lookups.
        type = GraphQLID;
    }

    if (!field.config.nillable) {
        type = new GraphQLNonNull(type);
    }

    return {
        type,
        resolve,
        extensions: {
            sfdc: field,
        },
    };
}

function createQuery(state: State, entities: Entity[]): GraphQLObjectType {
    return new GraphQLObjectType({
        name: 'Query',
        fields: () => {
            return Object.assign(
                {},
                ...entities.map((entity) => createEntityQueries(state, entity)),
            );
        },
    });
}

function createEntityQueries(
    state: State,
    entity: Entity,
): Record<string, GraphQLFieldConfig<unknown, unknown>> {
    const { gqlName, sfdcName } = entity;
    const { resolvers } = state.config;

    const type = state.types.get(sfdcName)!;

    const orderByInputType = new GraphQLList(
        new GraphQLInputObjectType({
            name: `${sfdcName}OrderBy`,
            fields: Object.fromEntries(
                entity.fields
                    .filter((field) => field.config.sortable)
                    .map((field) => [
                        field.gqlName,
                        {
                            type: isReferenceField(field)
                                ? state.orderByTypes.get(field.sfdcReferencedEntityName) ??
                                  ORDER_BY_ENUM_TYPE
                                : ORDER_BY_ENUM_TYPE,
                        },
                    ]),
            ),
        }),
    );
    state.orderByTypes.set(sfdcName, orderByInputType);

    const columnExprInputType: GraphQLInputObjectType = new GraphQLInputObjectType({
        name: `${sfdcName}ColumnExpr`,
        fields: () => {
            return {
                _and: {
                    type: new GraphQLList(columnExprInputType),
                },
                _or: {
                    type: new GraphQLList(columnExprInputType),
                },
                ...Object.fromEntries(
                    entity.fields
                        .filter(
                            (field): field is ScalarField | ReferenceField =>
                                field.config.filterable && !isPolymorphicReference(field),
                        )
                        .map((field) => [
                            field.gqlName,
                            {
                                type: isReferenceField(field)
                                    ? state.columnExprTypes.get(field.sfdcReferencedEntityName) ??
                                      ORDER_BY_ENUM_TYPE
                                    : OPERATOR_INPUT_TYPES_MAPPING[field.type],
                            },
                        ]),
                ),
            };
        },
    });
    state.columnExprTypes.set(sfdcName, columnExprInputType);

    return {
        [gqlName]: {
            args: {
                ...PAGINATION_INPUT_ARGS,
                where: {
                    type: columnExprInputType,
                },
                order_by: {
                    type: orderByInputType,
                },
            },
            type: new GraphQLList(type),
            resolve: resolvers?.queryMany?.(entity),
        },
        [`${gqlName}_by_id`]: {
            args: BY_ID_INPUT_ARGS,
            type: type,
            resolve: resolvers?.query?.(entity),
        },
    };
}

function createInputOperator({ type }: { type: GraphQLScalarType }): GraphQLInputObjectType {
    const { name } = type;
    const fields: Record<string, GraphQLInputFieldConfig> = {
        _eq: { type },
        _neq: { type },
        _gt: { type },
        _lt: { type },
        _gte: { type },
        _lte: { type },
        _in: { type: new GraphQLList(type) },
        _nin: { type: new GraphQLList(type) },
    };

    if (name === 'MultiPicklist') {
        fields._includes = { type };
        fields._excludes = { type };
    } else if (name === 'Text') {
        fields._like = { type };
    }

    return new GraphQLInputObjectType({
        name: `${name}Operator`,
        fields,
    });
}
