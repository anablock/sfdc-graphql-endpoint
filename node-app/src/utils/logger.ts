export interface LoggerFn {
    (...args: unknown[]): void;
}

export interface Logger {
    info: LoggerFn;
    warn: LoggerFn;
    error: LoggerFn;
    debug: LoggerFn;
}
