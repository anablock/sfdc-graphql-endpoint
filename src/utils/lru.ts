interface ListNode<K, V> {
    key: K;
    value: V;
    prev: ListNode<K, V> | null;
    next: ListNode<K, V> | null;
}

export class LRU<K, V> {
    limit: number;
    #map: Map<K, ListNode<K, V>> = new Map();
    #head: ListNode<K, V> | null = null;
    #tail: ListNode<K, V> | null = null;

    constructor(limit = 100) {
        this.limit = limit;
    }

    get(key: K): V | null {
        const entry = this.#map.get(key);

        if (entry === undefined) {
            return null;
        }

        if (entry !== this.#head) {
            this.set(key, entry.value);
        }

        return entry.value;
    }

    set(key: K, value: V): void {
        const existingEntry = this.#map.get(key);

        if (existingEntry !== undefined) {
            this.delete(key);
        } else if (this.#map.size >= this.limit && this.#tail !== null) {
            this.delete(this.#tail.key);
        }

        const newEntry = {
            key,
            value,
            prev: null,
            next: this.#head,
        };

        if (this.#head === null) {
            this.#head = this.#tail = newEntry;
        } else {
            this.#head.prev = newEntry;
            this.#head = newEntry;
        }

        this.#map.set(key, newEntry);
    }

    delete(key: K): void {
        const entry = this.#map.get(key);

        if (entry === undefined) {
            return;
        }

        this.#map.delete(key);

        if (entry.prev !== null) {
            entry.prev.next = entry.next;
        } else {
            this.#head = entry.next;
        }

        if (entry.next !== null) {
            entry.next.prev = entry.prev;
        } else {
            this.#tail = entry;
        }
    }

    clear() {
        this.#head = this.#tail = null;
        this.#map.clear();
    }

    *keys(): Generator<K> {
        const node = this.#head;

        while (node !== null) {
            yield node.key;
        }
    }

    *values(): Generator<V> {
        const node = this.#head;

        while (node !== null) {
            yield node.value;
        }
    }

    *entries(): Generator<[K, V]> {
        const node = this.#head;

        while (node !== null) {
            yield [node.key, node.value];
        }
    }
}
