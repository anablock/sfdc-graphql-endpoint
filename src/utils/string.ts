export function camelCase(str: string): string {
    return str[0].toLowerCase() + str.slice(1);
}
