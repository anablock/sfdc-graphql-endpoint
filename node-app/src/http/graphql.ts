import * as graphql from 'graphql';
import { GraphQLSchema, Source, DocumentNode } from 'graphql';
import { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';

import { SchemaBuilder } from '../schema-builder.js';
import { Graph } from '../graph.js';

import { LRU } from '../utils/lru.js';

import { Api } from '../sfdc/api.js';

import sfdcFastifyPlugin from './sfdc-plugin.js';

const ENTITIES = ['Account', 'User', 'Lead', 'Opportunity'];

interface GraphQLQueryString {
    query?: string;
    variables?: string;
    operationName?: string;
}

interface GraphQLParams {
    query?: string;
    variables?: Record<string, unknown>;
    operationName?: string;
}

export async function graphqlFastifyPlugin(fastify: FastifyInstance) {
    let schema: GraphQLSchema | undefined;
    const queryCache = new LRU<string, DocumentNode>(100);

    fastify.register(sfdcFastifyPlugin);

    fastify.get<{
        Params: GraphQLQueryString;
    }>('/graphql', async (request, response) => {
        if (request.headers.accept?.includes('text/html')) {
            return respondWithGraphIql(response);
        }

        let variables;
        try {
            variables = request.params.variables ? JSON.parse(request.params.variables) : {};
        } catch (error) {
            throw {
                statusCode: 400,
                message: 'Invalid GraphQL query.',
                errors: ['Failed to parse variables'],
            };
        }

        return executeQuery(request, {
            ...request.params,
            variables,
        });
    });

    fastify.post<{
        Body: GraphQLParams;
    }>('/graphql', async (request, response) => {
        return executeQuery(request, {
            ...request.body,
        });
    });

    async function executeQuery(
        request: FastifyRequest,
        params: GraphQLParams,
    ): Promise<graphql.ExecutionResult> {
        const { api, connection } = fastify.sfdc;
        const { query, variables, operationName } = params;

        if (!schema) {
            request.log.info('Fetching graphQL schema');
            schema = await buildSchema(api);
        }

        if (!query) {
            throw {
                statusCode: 400,
                message: 'Invalid GraphQL query.',
                errors: ['Missing query body'],
            };
        }

        const queryDocumentAst = parseAndValidateQuery(schema, query);

        return graphql.execute({
            schema,
            operationName,
            variableValues: variables,
            document: queryDocumentAst,
            contextValue: {
                api,
                connection,
                logger: request.log
            },
        });
    }

    function parseAndValidateQuery(schema: GraphQLSchema, query: string): DocumentNode {
        let documentAst = queryCache.get(query);

        if (documentAst === null) {
            try {
                documentAst = graphql.parse(new Source(query));
            } catch (err: unknown) {
                throw { statusCode: 400, message: 'GraphQL syntax error.', errors: [err] };
            }

            const validationErrors = graphql.validate(schema, documentAst);
            if (validationErrors.length > 0) {
                throw {
                    statusCode: 400,
                    message: 'Invalid GraphQL query.',
                    errors: validationErrors,
                };
            }

            queryCache.set(query, documentAst);
        }

        return documentAst;
    }
}

async function buildSchema(api: Api): Promise<GraphQLSchema> {
    const sObjects = await Promise.all(ENTITIES.map((entity) => api.describeSObject(entity)));

    const graph = new Graph(sObjects);
    const schemaBuilder = new SchemaBuilder(graph);

    const schema = schemaBuilder.buildSchema();
    graphql.assertValidSchema(schema);

    return schema;
}

function respondWithGraphIql(response: FastifyReply) {
    response.type('text/html').send(`
        <html>
            <head>
                <title>Simple GraphiQL Example</title>
                <link href="https://unpkg.com/graphiql/graphiql.min.css" rel="stylesheet" />
            </head>
            <body style="margin: 0;">
                <div id="graphiql" style="height: 100vh;"></div>

                <script
                    crossorigin
                    src="https://unpkg.com/react/umd/react.production.min.js"
                ></script>
                <script
                    crossorigin
                    src="https://unpkg.com/react-dom/umd/react-dom.production.min.js"
                ></script>
                <script
                    crossorigin
                    src="https://unpkg.com/graphiql/graphiql.min.js"
                ></script>

                <script>
                    const fetcher = GraphiQL.createFetcher({ url: '/graphql' });

                    ReactDOM.render(
                        React.createElement(GraphiQL, { fetcher: fetcher }),
                        document.getElementById('graphiql'),
                    );
                </script>
            </body>
        </html>
    `);
}
