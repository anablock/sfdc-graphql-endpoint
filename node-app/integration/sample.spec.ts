import { buildClientSchema, getIntrospectionQuery, printSchema } from 'graphql';

import getApp from '../src/http/app.js';

import { gql } from './utils.js';

const instanceUrl = process.env.INSTANCE_URL;
const accessToken = process.env.ACCESS_TOKEN;

if (!instanceUrl || !accessToken) {
    let msg = `INSTANCE_URL and ACCESS_TOKEN environment variables must be set to run the integration tests.\n`;
    msg += `Run "sfdx force:user:display" from the "integration/sfdx-org" directory to retrieve those informations.`;

    throw new Error(msg);
}

describe('Sample entity', () => {
    const app = getApp({
        instanceUrl,
        accessToken,
        entities: ['Sample__c'],
    });

    async function executeQuery({ query }: { query: string }): Promise<any> {
        const response = await app.inject({
            method: 'POST',
            url: '/graphql',
            payload: {
                query,
            },
        });

        expect(response.statusCode).toBe(200);
        return response.json();
    }

    test('introspect schema', async () => {
        const { data } = await executeQuery({
            query: getIntrospectionQuery({
                schemaDescription: true,
            }),
        });

        const schema = buildClientSchema(data);
        expect(printSchema(schema)).toMatchSnapshot();
    });

    test('retrieve all the sample records', async () => {
        const { data } = await executeQuery({
            query: gql`
                {
                    Sample__c(limit: 100) {
                        id
                        name
                    }
                }
            `,
        });

        expect(data).toMatchObject({
            Sample__c: expect.any(Array),
        });

        for (const record of data.Sample__c) {
            expect(record).toMatchObject({
                id: expect.any(String),
                name: expect.any(String),
            });
        }
    });

    test('retrieve 2 records', async () => {
        const { data } = await executeQuery({
            query: gql`
                {
                    Sample__c(limit: 2) {
                        name
                    }
                }
            `,
        });

        expect(data.Sample__c).toHaveLength(2);
    });

    test('retrieve records ordered by name ASC', async () => {
        const { data } = await executeQuery({
            query: gql`
                {
                    Sample__c(limit: 10, order_by: { name: ASC }) {
                        name
                    }
                }
            `,
        });

        const names: string[] = data.Sample__c.map((record: any) => record.name);
        expect(names).toEqual([...names].sort());
    });

    test('retrieve records ordered by name DESC', async () => {
        const { data } = await executeQuery({
            query: gql`
                {
                    Sample__c(limit: 10, order_by: { name: DESC }) {
                        name
                    }
                }
            `,
        });

        const names: string[] = data.Sample__c.map((record: any) => record.name);
        expect(names).toEqual([...names].sort((a, b) => b.localeCompare(a)));
    });

    // TODO: Fix me, something is wrong with the offset.
    test.skip('retrieve next record using offset', async () => {
        const fetchAllRes = await executeQuery({
            query: gql`
                {
                    Sample__c(limit: 4) {
                        id
                        name
                    }
                }
            `,
        });
        const [firstRecordId, nextRecordId] = fetchAllRes.data.Sample__c.map(
            (record: any) => record.id,
        );

        const fetchFirstRes = await executeQuery({
            query: gql`
                {
                    Sample__c(limit: 1, offset: 0) {
                        id
                        name
                    }
                }
            `,
        });
        expect(fetchFirstRes.data.Sample__c[0].id).toBe(firstRecordId);

        const fetchNextRes = await executeQuery({
            query: gql`
                {
                    Sample__c(limit: 1, offset: 0) {
                        id
                        name
                    }
                }
            `,
        });
        expect(fetchNextRes.data.Sample__c[0].id).toBe(nextRecordId);
    });
});
