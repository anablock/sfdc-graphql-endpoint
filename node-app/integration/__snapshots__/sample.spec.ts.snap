// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sample entity introspect schema 1`] = `
"type Sample__c {
  id: ID!
  owner: ID!
  isDeleted: Boolean!
  name: String
  createdDate: DateTime!
  createdBy: ID!
  lastModifiedDate: DateTime!
  lastModifiedBy: ID!
  systemModstamp: DateTime!
  lastViewedDate: DateTime
  lastReferencedDate: DateTime
  sampleCheckbox__c: Boolean!
  sampleCurrency__c: Currency
  sampleDate__c: Date
  sampleDateTime__c: DateTime
  sampleTime__c: Time
  sampleEmail__c: Email
  sampleNumber__c: Float
  samplePercent__c: Percent
  samplePhone__c: Phone
  samplePicklist__c: Picklist
  sampleMultiPicklist__c: MultiPicklist
  sampleText__c: String
  sampleEncryptedText__c: EncryptedString
  sampleTextArea__c: TextArea
  sampleLongTextArea__c: TextArea
  sampleRichTextArea__c: TextArea
  sampleURL__c: URL
}

scalar DateTime

scalar Currency

scalar Date

scalar Time

scalar Email

scalar Percent

scalar Phone

scalar Picklist

scalar MultiPicklist

scalar EncryptedString

scalar TextArea

scalar URL

type Query {
  Sample__c(limit: Int!, offset: Int, where: Sample__cColumnExpr, order_by: [Sample__cOrderBy]): [Sample__c]
  Sample__c_by_id(id: ID): Sample__c
}

input Sample__cColumnExpr {
  _and: [Sample__cColumnExpr]
  _or: [Sample__cColumnExpr]
  id: IDOperator
  isDeleted: BooleanOperator
  name: StringOperator
  createdDate: DateTimeOperator
  createdBy: OrderByEnum
  lastModifiedDate: DateTimeOperator
  lastModifiedBy: OrderByEnum
  systemModstamp: DateTimeOperator
  lastViewedDate: DateTimeOperator
  lastReferencedDate: DateTimeOperator
  sampleCheckbox__c: BooleanOperator
  sampleCurrency__c: CurrencyOperator
  sampleDate__c: DateOperator
  sampleDateTime__c: DateTimeOperator
  sampleTime__c: TimeOperator
  sampleEmail__c: EmailOperator
  sampleNumber__c: FloatOperator
  samplePercent__c: PercentOperator
  samplePhone__c: PhoneOperator
  samplePicklist__c: PicklistOperator
  sampleMultiPicklist__c: MultiPicklistOperator
  sampleText__c: StringOperator
  sampleTextArea__c: TextAreaOperator
  sampleURL__c: URLOperator
}

input IDOperator {
  _eq: ID
  _neq: ID
  _gt: ID
  _lt: ID
  _gte: ID
  _lte: ID
  _in: [ID]
  _nin: [ID]
}

input BooleanOperator {
  _eq: Boolean
  _neq: Boolean
  _gt: Boolean
  _lt: Boolean
  _gte: Boolean
  _lte: Boolean
  _in: [Boolean]
  _nin: [Boolean]
}

input StringOperator {
  _eq: String
  _neq: String
  _gt: String
  _lt: String
  _gte: String
  _lte: String
  _in: [String]
  _nin: [String]
}

input DateTimeOperator {
  _eq: DateTime
  _neq: DateTime
  _gt: DateTime
  _lt: DateTime
  _gte: DateTime
  _lte: DateTime
  _in: [DateTime]
  _nin: [DateTime]
}

enum OrderByEnum {
  ASC
  DESC
  ASC_NULLS_FIRST
  ASC_NULLS_LAST
  DESC_NULLS_FIRST
  DESC_NULLS_LAST
}

input CurrencyOperator {
  _eq: Currency
  _neq: Currency
  _gt: Currency
  _lt: Currency
  _gte: Currency
  _lte: Currency
  _in: [Currency]
  _nin: [Currency]
}

input DateOperator {
  _eq: Date
  _neq: Date
  _gt: Date
  _lt: Date
  _gte: Date
  _lte: Date
  _in: [Date]
  _nin: [Date]
}

input TimeOperator {
  _eq: Time
  _neq: Time
  _gt: Time
  _lt: Time
  _gte: Time
  _lte: Time
  _in: [Time]
  _nin: [Time]
}

input EmailOperator {
  _eq: Email
  _neq: Email
  _gt: Email
  _lt: Email
  _gte: Email
  _lte: Email
  _in: [Email]
  _nin: [Email]
}

input FloatOperator {
  _eq: Float
  _neq: Float
  _gt: Float
  _lt: Float
  _gte: Float
  _lte: Float
  _in: [Float]
  _nin: [Float]
}

input PercentOperator {
  _eq: Percent
  _neq: Percent
  _gt: Percent
  _lt: Percent
  _gte: Percent
  _lte: Percent
  _in: [Percent]
  _nin: [Percent]
}

input PhoneOperator {
  _eq: Phone
  _neq: Phone
  _gt: Phone
  _lt: Phone
  _gte: Phone
  _lte: Phone
  _in: [Phone]
  _nin: [Phone]
}

input PicklistOperator {
  _eq: Picklist
  _neq: Picklist
  _gt: Picklist
  _lt: Picklist
  _gte: Picklist
  _lte: Picklist
  _in: [Picklist]
  _nin: [Picklist]
}

input MultiPicklistOperator {
  _eq: MultiPicklist
  _neq: MultiPicklist
  _gt: MultiPicklist
  _lt: MultiPicklist
  _gte: MultiPicklist
  _lte: MultiPicklist
  _in: [MultiPicklist]
  _nin: [MultiPicklist]
  _includes: MultiPicklist
  _excludes: MultiPicklist
}

input TextAreaOperator {
  _eq: TextArea
  _neq: TextArea
  _gt: TextArea
  _lt: TextArea
  _gte: TextArea
  _lte: TextArea
  _in: [TextArea]
  _nin: [TextArea]
}

input URLOperator {
  _eq: URL
  _neq: URL
  _gt: URL
  _lt: URL
  _gte: URL
  _lte: URL
  _in: [URL]
  _nin: [URL]
}

input Sample__cOrderBy {
  id: OrderByEnum
  owner: OrderByEnum
  isDeleted: OrderByEnum
  name: OrderByEnum
  createdDate: OrderByEnum
  createdBy: OrderByEnum
  lastModifiedDate: OrderByEnum
  lastModifiedBy: OrderByEnum
  systemModstamp: OrderByEnum
  lastViewedDate: OrderByEnum
  lastReferencedDate: OrderByEnum
  sampleCheckbox__c: OrderByEnum
  sampleCurrency__c: OrderByEnum
  sampleDate__c: OrderByEnum
  sampleDateTime__c: OrderByEnum
  sampleTime__c: OrderByEnum
  sampleEmail__c: OrderByEnum
  sampleNumber__c: OrderByEnum
  samplePercent__c: OrderByEnum
  samplePhone__c: OrderByEnum
  samplePicklist__c: OrderByEnum
  sampleText__c: OrderByEnum
  sampleTextArea__c: OrderByEnum
  sampleURL__c: OrderByEnum
}"
`;
